File: android/src/main/java/net/no_mad/tts/TextToSpeechModule.java
package net.no_mad.tts;

import android.media.AudioManager;
import android.os.Build;
import android.os.Bundle;
import android.content.Intent;
import android.content.ActivityNotFoundException;
import android.app.Activity;
import android.net.Uri;
import android.speech.tts.TextToSpeech;
import android.speech.tts.UtteranceProgressListener;
import android.speech.tts.Voice;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.PackageManager.NameNotFoundException;
import com.facebook.react.bridge.*;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

public class TextToSpeechModule extends ReactContextBaseJavaModule {

    private TextToSpeech tts;
    private Boolean ready = null;
    private ArrayList<Promise> initStatusPromises;

    private boolean ducking = false;
    private AudioManager audioManager;
    private AudioManager.OnAudioFocusChangeListener afChangeListener;

    private Map<String, Locale> localeCountryMap;
    private Map<String, Locale> localeLanguageMap;

    public TextToSpeechModule(ReactApplicationContext reactContext) {
        super(reactContext);
        audioManager = (AudioManager) reactContext.getApplicationContext().getSystemService(reactContext.AUDIO_SERVICE);
        initStatusPromises = new ArrayList<Promise>();
        //initialize ISO3, ISO2 languague country code mapping.
        initCountryLanguageCodeMapping();

        tts = new TextToSpeech(getReactApplicationContext(), new TextToSpeech.OnInitListener() {
            @Override
            public void onInit(int status) {
                synchronized(initStatusPromises) {
                    ready = (status == TextToSpeech.SUCCESS) ? Boolean.TRUE : Boolean.FALSE;
                    for(Promise p: initStatusPromises) {
                        resolveReadyPromise(p);
                    }
                    initStatusPromises.clear();
                }
            }
        });

        setUtteranceProgress();
    }

    private void setUtteranceProgress() {
        if(tts != null)
        {
            tts.setOnUtteranceProgressListener(new UtteranceProgressListener() {
                @Override
                public void onStart(String utteranceId) {
                    sendEvent("tts-start", utteranceId);
                }

                @Override
                public void onDone(String utteranceId) {
                    if(ducking) {
                        audioManager.abandonAudioFocus(afChangeListener);
                    }
                    sendEvent("tts-finish", utteranceId);
                }

                @Override
                public void onError(String utteranceId) {
                    if(ducking) {
                        audioManager.abandonAudioFocus(afChangeListener);
                    }
                    sendEvent("tts-error", utteranceId);
                }

                @Override
                public void onStop(String utteranceId, boolean interrupted) {
                    if(ducking) {
                        audioManager.abandonAudioFocus(afChangeListener);
                    }
                    sendEvent("tts-cancel", utteranceId);
                }

                @Override
                public void onRangeStart (String utteranceId, int start, int end, int frame) {
                    WritableMap params = Arguments.createMap();
                    params.putString("utteranceId", utteranceId);
                    params.putInt("start", start);
                    params.putInt("end", end);
                    params.putInt("frame", frame);
                    sendEvent("tts-progress", params);
                }
            });
        }
    }

    private void initCountryLanguageCodeMapping() {
        String[] countries = Locale.getISOCountries();
        localeCountryMap = new HashMap<String, Locale>(countries.length);
        for (String country: countries) {
            Locale locale = new Locale("", country);
            localeCountryMap.put(locale.getISO3Country().toUpperCase(), locale);
        }
        String[] languages = Locale.getISOLanguages();
        localeLanguageMap = new HashMap<String, Locale>(languages.length);
        for (String language: languages) {
            Locale locale = new Locale(language);
            localeLanguageMap.put(locale.getISO3Language(), locale);
        }
    }

    private String iso3CountryCodeToIso2CountryCode(String iso3CountryCode) {
        return localeCountryMap.get(iso3CountryCode).getCountry();
    }

    private String iso3LanguageCodeToIso2LanguageCode(String iso3LanguageCode) {
        return localeLanguageMap.get(iso3LanguageCode).getLanguage();
    }

    private void resolveReadyPromise(Promise promise) {
        if (ready == Boolean.TRUE) {
            promise.resolve("success");
        }
        else {
            promise.reject("no_engine", "No TTS engine installed");
        }
    }

    private static void resolvePromiseWithStatusCode(int statusCode, Promise promise) {
        switch (statusCode) {
            case TextToSpeech.SUCCESS:
                promise.resolve("success");
                break;
            case TextToSpeech.LANG_COUNTRY_AVAILABLE:
                promise.resolve("lang_country_available");
                break;
            case TextToSpeech.LANG_COUNTRY_VAR_AVAILABLE:
                promise.resolve("lang_country_var_available");
                break;
            case TextToSpeech.ERROR_INVALID_REQUEST:
                promise.reject("invalid_request", "Failure caused by an invalid request");
                break;
            case TextToSpeech.ERROR_NETWORK:
                promise.reject("network_error", "Failure caused by a network connectivity problems");
                break;
            case TextToSpeech.ERROR_NETWORK_TIMEOUT:
                promise.reject("network_timeout", "Failure caused by network timeout.");
                break;
            case TextToSpeech.ERROR_NOT_INSTALLED_YET:
                promise.reject("not_installed_yet", "Unfinished download of voice data");
                break;
            case TextToSpeech.ERROR_OUTPUT:
                promise.reject("output_error", "Failure related to the output (audio device or a file)");
                break;
            case TextToSpeech.ERROR_SERVICE:
                promise.reject("service_error", "Failure of a TTS service");
                break;
            case TextToSpeech.ERROR_SYNTHESIS:
                promise.reject("synthesis_error", "Failure of a TTS engine to synthesize the given input");
                break;
            case TextToSpeech.LANG_MISSING_DATA:
                promise.reject("lang_missing_data", "Language data is missing");
                break;
            case TextToSpeech.LANG_NOT_SUPPORTED:
                promise.reject("lang_not_supported", "Language is not supported");
                break;
            default:
                promise.reject("error", "Unknown error code: " + statusCode);
                break;
          }
    }

    private boolean isPackageInstalled(String packageName) {
        PackageManager pm = getReactApplicationContext().getPackageManager();
        try {
            PackageInfo pi = pm.getPackageInfo(packageName, 0);
            return true;
        } catch (NameNotFoundException e) {
            return false;
        }
    }

    @Override
    public String getName() {
        return "TextToSpeech";
    }

    @ReactMethod
    public void getInitStatus(Promise promise) {
        synchronized(initStatusPromises) {
            if(ready == null) {
                initStatusPromises.add(promise);
            } else {
                resolveReadyPromise(promise);
            }
        }
    }

    @ReactMethod
    public void speak(String utterance, ReadableMap params, Promise promise) {
        if(notReady(promise)) return;

        if(ducking) {
            // Request audio focus for playback
            int amResult = audioManager.requestAudioFocus(afChangeListener,
                                                          // Use the music stream.
                                                          AudioManager.STREAM_MUSIC,
                                                          // Request permanent focus.
                                                          AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK);

            if(amResult != AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
                promise.reject("Android AudioManager error, failed to request audio focus");
                return;
            }
        }

        String utteranceId = Integer.toString(utterance.hashCode());

        int speakResult = speak(utterance, utteranceId, params);
        if(speakResult == TextToSpeech.SUCCESS) {
            promise.resolve(utteranceId);
        } else {
            resolvePromiseWithStatusCode(speakResult, promise);
        }
    }

    @ReactMethod
    public void setDefaultLanguage(String language, Promise promise) {
        if(notReady(promise)) return;

        Locale locale = null;

        if(language.indexOf("-") != -1) {
            String[] parts = language.split("-");
            locale = new Locale(parts[0], parts[1]);
        } else {
            locale = new Locale(language);
        }

        try {
          int result = tts.setLanguage(locale);
          resolvePromiseWithStatusCode(result, promise);
        } catch (Exception e) {
          promise.reject("error", "Unknown error code");
        }
    }

    @ReactMethod
    public void setDucking(Boolean ducking, Promise promise) {
        if(notReady(promise)) return;
        this.ducking = ducking;
        promise.resolve("success");
    }

    @ReactMethod
    public void setDefaultRate(Float rate, Boolean skipTransform, Promise promise) {
        if(notReady(promise)) return;

        if(skipTransform) {
            int result = tts.setSpeechRate(rate);
            resolvePromiseWithStatusCode(result, promise);
        } else {
            // normalize android rate
            // rate value will be in the range 0.0 to 1.0
            // let's convert it to the range of values Android platform expects,
            // where 1.0 is no change of rate and 2.0 is the twice faster rate
            float androidRate = rate.floatValue() < 0.5f ?
                    rate.floatValue() * 2 : // linear fit {0, 0}, {0.25, 0.5}, {0.5, 1}
                    rate.floatValue() * 4 - 1; // linear fit {{0.5, 1}, {0.75, 2}, {1, 3}}
            int result = tts.setSpeechRate(androidRate);
            resolvePromiseWithStatusCode(result, promise);
        }
    }

    @ReactMethod
    public void setDefaultPitch(Float pitch, Promise promise) {
        if(notReady(promise)) return;
        int result = tts.setPitch(pitch);
        resolvePromiseWithStatusCode(result, promise);
    }

    @ReactMethod
    public void setDefaultVoice(String voiceId, Promise promise) {
        if(notReady(promise)) return;

        if (Build.VERSION.SDK_INT >= 21) {
            try {
                for(Voice voice: tts.getVoices()) {
                    if(voice.getName().equals(voiceId)) {
                        int result = tts.setVoice(voice);
                        resolvePromiseWithStatusCode(result, promise);
                        return;
                    }
                }
            } catch (Exception e) {
              // Purposefully ignore exceptions here due to some buggy TTS engines.
              // See http://stackoverflow.com/questions/26730082/illegalargumentexception-invalid-int-os-with-samsung-tts
            }
            promise.reject("not_found", "The selected voice was not found");
        } else {
            promise.reject("not_available", "Android API 21 level or higher is required");
        }
    }

    @ReactMethod
    public void voices(Promise promise) {
        if(notReady(promise)) return;

        WritableArray voiceArray = Arguments.createArray();

        if (Build.VERSION.SDK_INT >= 21) {
            try {
                for(Voice voice: tts.getVoices()) {
                    WritableMap voiceMap = Arguments.createMap();
                    voiceMap.putString("id", voice.getName());
                    voiceMap.putString("name", voice.getName());

                    String language = iso3LanguageCodeToIso2LanguageCode(voice.getLocale().getISO3Language());
                    String country = voice.getLocale().getISO3Country();
                    if(country != "") {
                        language += "-" + iso3CountryCodeToIso2CountryCode(country);
                    }

                    voiceMap.putString("language", language);
                    voiceMap.putInt("quality", voice.getQuality());
                    voiceMap.putInt("latency", voice.getLatency());
                    voiceMap.putBoolean("networkConnectionRequired", voice.isNetworkConnectionRequired());
                    voiceMap.putBoolean("notInstalled", voice.getFeatures().contains(TextToSpeech.Engine.KEY_FEATURE_NOT_INSTALLED));
                    voiceArray.pushMap(voiceMap);
                }
            } catch (Exception e) {
              // Purposefully ignore exceptions here due to some buggy TTS engines.
              // See http://stackoverflow.com/questions/26730082/illegalargumentexception-invalid-int-os-with-samsung-tts
            }
        }

        promise.resolve(voiceArray);
    }

    @ReactMethod
    public void setDefaultEngine(String engineName, final Promise promise) {
        if(notReady(promise)) return;

        if(isPackageInstalled(engineName)) {
            ready = null;
            onCatalystInstanceDestroy();
            tts = new TextToSpeech(getReactApplicationContext(), new TextToSpeech.OnInitListener() {
                @Override
                public void onInit(int status) {
                    synchronized(initStatusPromises) {
                        ready = (status == TextToSpeech.SUCCESS) ? Boolean.TRUE : Boolean.FALSE;
                        for(Promise p: initStatusPromises) {
                            resolveReadyPromise(p);
                        }
                        initStatusPromises.clear();
                        promise.resolve(ready);
                    }
                }
            }, engineName);

            setUtteranceProgress();
        } else {
            promise.reject("not_found", "The selected engine was not found");
        }
    }

    @ReactMethod
    public void engines(Promise promise) {
        if(notReady(promise)) return;

        WritableArray engineArray = Arguments.createArray();

        if (Build.VERSION.SDK_INT >= 14) {
            try {
                String defaultEngineName = tts.getDefaultEngine();
                for(TextToSpeech.EngineInfo engine: tts.getEngines()) {
                    WritableMap engineMap = Arguments.createMap();

                    engineMap.putString("name", engine.name);
                    engineMap.putString("label", engine.label);
                    engineMap.putBoolean("default", engine.name.equals(defaultEngineName));
                    engineMap.putInt("icon", engine.icon);

                    engineArray.pushMap(engineMap);
                }
            } catch (Exception e) {
                promise.reject("error", "Unknown error code");
            }
        }

        promise.resolve(engineArray);
    }

    @ReactMethod
    public void stop(Promise promise) {
        if(notReady(promise)) return;

        int result = tts.stop();
        boolean resultValue = (result == TextToSpeech.SUCCESS) ? Boolean.TRUE : Boolean.FALSE;
        promise.resolve(resultValue);
    }

    @ReactMethod
    private void requestInstallEngine(Promise promise) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse("market://details?id=com.google.android.tts"));
        try {
            getCurrentActivity().startActivity(intent);
            promise.resolve("success");
        } catch (Exception e) {
            promise.reject("error", "Could not open Google Text to Speech App in the Play Store");
        }
    }

    @ReactMethod
    private void requestInstallData(Promise promise) {
        Intent intent = new Intent();
        intent.setAction(TextToSpeech.Engine.ACTION_INSTALL_TTS_DATA);
        try {
            getCurrentActivity().startActivity(intent);
            promise.resolve("success");
        } catch (ActivityNotFoundException e) {
            promise.reject("no_engine", "No TTS engine installed");
        }
    }

    /**
     * called on React Native Reloading JavaScript
     * https://stackoverflow.com/questions/15563361/tts-leaked-serviceconnection
     */
    @Override
    public void onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy();
        if(tts != null) {
            tts.stop();
            tts.shutdown();
        }
    }

    private boolean notReady(Promise promise) {
        if(ready == null) {
            promise.reject("not_ready", "TTS is not ready");
            return true;
        }
        else if(ready != Boolean.TRUE) {
            resolveReadyPromise(promise);
            return true;
        }
        return false;
    }

    @SuppressWarnings("deprecation")
    private int speak(String utterance, String utteranceId, ReadableMap inputParams) {
        String audioStreamTypeString = inputParams.hasKey("KEY_PARAM_STREAM") ? inputParams.getString("KEY_PARAM_STREAM") : "";
        float volume = inputParams.hasKey("KEY_PARAM_VOLUME") ? (float) inputParams.getDouble("KEY_PARAM_VOLUME") : 1.0f;
        float pan = inputParams.hasKey("KEY_PARAM_PAN") ? (float) inputParams.getDouble("KEY_PARAM_PAN") : 0.0f;

        int audioStreamType;
        switch(audioStreamTypeString) {
            /*
            // This has been added in API level 26, commenting out for now

            case "STREAM_ACCESSIBILITY":
                audioStreamType = AudioManager.STREAM_ACCESSIBILITY;
                break;
            */
            case "STREAM_ALARM":
                audioStreamType = AudioManager.STREAM_ALARM;
                break;
            case "STREAM_DTMF":
                audioStreamType = AudioManager.STREAM_DTMF;
                break;
            case "STREAM_MUSIC":
                audioStreamType = AudioManager.STREAM_MUSIC;
                break;
            case "STREAM_NOTIFICATION":
                audioStreamType = AudioManager.STREAM_NOTIFICATION;
                break;
            case "STREAM_RING":
                audioStreamType = AudioManager.STREAM_RING;
                break;
            case "STREAM_SYSTEM":
                audioStreamType = AudioManager.STREAM_SYSTEM;
                break;
            case "STREAM_VOICE_CALL":
                audioStreamType = AudioManager.STREAM_VOICE_CALL;
                break;
            default:
                audioStreamType = AudioManager.USE_DEFAULT_STREAM_TYPE;
        }

        if (Build.VERSION.SDK_INT >= 21) {
            Bundle params = new Bundle();
            params.putInt(TextToSpeech.Engine.KEY_PARAM_STREAM, audioStreamType);
            params.putFloat(TextToSpeech.Engine.KEY_PARAM_VOLUME, volume);
            params.putFloat(TextToSpeech.Engine.KEY_PARAM_PAN, pan);
            return tts.speak(utterance, TextToSpeech.QUEUE_ADD, params, utteranceId);
        } else {
            HashMap<String, String> params = new HashMap();
            params.put(TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID, utteranceId);
            params.put(TextToSpeech.Engine.KEY_PARAM_STREAM, String.valueOf(audioStreamType));
            params.put(TextToSpeech.Engine.KEY_PARAM_VOLUME, String.valueOf(volume));
            params.put(TextToSpeech.Engine.KEY_PARAM_PAN, String.valueOf(pan));
            return tts.speak(utterance, TextToSpeech.QUEUE_ADD, params);
        }
    }

    private void sendEvent(String eventName, String utteranceId) {
        WritableMap params = Arguments.createMap();
        params.putString("utteranceId", utteranceId);
        sendEvent(eventName, params);
    }

    private void sendEvent(String eventName, WritableMap params) {
        getReactApplicationContext()
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
    }
    
    @ReactMethod
    public void removeListeners(Integer count) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    @ReactMethod
    public void addListener(String eventName) {
        // Keep: Required for RN built in Event Emitter Calls.
    }
}

File: android/src/main/java/net/no_mad/tts/TextToSpeechModule.kt
package net.no_mad.tts

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Intent
import android.media.AudioManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import android.speech.tts.Voice
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule

class TextToSpeechModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    private var tts: TextToSpeech? = null
    private var ready: Boolean? = null
    private var initStatusPromises: ArrayList<Promise> = ArrayList()
    private var ducking = false
    private val audioManager = reactContext.applicationContext.getSystemService(ReactApplicationContext.AUDIO_SERVICE) as AudioManager
    private lateinit var afChangeListener: AudioManager.OnAudioFocusChangeListener

    private var localeCountryMap: Map<String, Locale> = HashMap()
    private var localeLanguageMap: Map<String, Locale> = HashMap()

    init {
        initCountryLanguageCodeMapping()

        tts = TextToSpeech(reactContext) { status ->
            synchronized(initStatusPromises) {
                ready = (status == TextToSpeech.SUCCESS)
                initStatusPromises.forEach { resolveReadyPromise(it) }
                initStatusPromises.clear()
            }
        }

        setUtteranceProgress()
    }

    private fun setUtteranceProgress() {
        tts?.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
            override fun onStart(utteranceId: String) {
                sendEvent("tts-start", utteranceId)
            }

            override fun onDone(utteranceId: String) {
                if (ducking) {
                    audioManager.abandonAudioFocus(afChangeListener)
                }
                sendEvent("tts-finish", utteranceId)
            }

            override fun onError(utteranceId: String) {
                if (ducking) {
                    audioManager.abandonAudioFocus(afChangeListener)
                }
                sendEvent("tts-error", utteranceId)
            }

            override fun onStop(utteranceId: String, interrupted: Boolean) {
                if (ducking) {
                    audioManager.abandonAudioFocus(afChangeListener)
                }
                sendEvent("tts-cancel", utteranceId)
            }

            override fun onRangeStart(utteranceId: String, start: Int, end: Int, frame: Int) {
                val params = Arguments.createMap().apply {
                    putString("utteranceId", utteranceId)
                    putInt("start", start)
                    putInt("end", end)
                    putInt("frame", frame)
                }
                sendEvent("tts-progress", params)
            }
        })
    }

    private fun initCountryLanguageCodeMapping() {
        Locale.getISOCountries().forEach { country ->
            localeCountryMap = localeCountryMap.plus(Pair(Locale("", country).isO3Country.toUpperCase(), Locale("", country)))
        }
        Locale.getISOLanguages().forEach { language ->
            localeLanguageMap = localeLanguageMap.plus(Pair(Locale(language).isO3Language, Locale(language)))
        }
    }

    private fun iso3CountryCodeToIso2CountryCode(iso3CountryCode: String): String {
        return localeCountryMap[iso3CountryCode]?.country ?: ""
    }

    private fun iso3LanguageCodeToIso2LanguageCode(iso3LanguageCode: String): String {
        return localeLanguageMap[iso3LanguageCode]?.language ?: ""
    }

    private fun resolveReadyPromise(promise: Promise) {
        if (ready == true) {
            promise.resolve("success")
        } else {
            promise.reject("no_engine", "No TTS engine installed")
        }
    }

    private fun resolvePromiseWithStatusCode(statusCode: Int, promise: Promise) {
        when (statusCode) {
            TextToSpeech.SUCCESS -> promise.resolve("success")
            TextToSpeech.LANG_COUNTRY_AVAILABLE -> promise.resolve("lang_country_available")
            TextToSpeech.LANG_COUNTRY_VAR_AVAILABLE -> promise.resolve("lang_country_var_available")
            TextToSpeech.ERROR_INVALID_REQUEST -> promise.reject("invalid_request", "Failure caused by an invalid request")
            TextToSpeech.ERROR_NETWORK -> promise.reject("network_error", "Failure caused by a network connectivity problems")
            TextToSpeech.ERROR_NETWORK_TIMEOUT -> promise.reject("network_timeout", "Failure caused by network timeout.")
            TextToSpeech.ERROR_NOT_INSTALLED_YET -> promise.reject("not_installed_yet", "Unfinished download of voice data")
            TextToSpeech.ERROR_OUTPUT -> promise.reject("output_error", "Failure related to the output (audio device or a file)")
            TextToSpeech.ERROR_SERVICE -> promise.reject("service_error", "Failure of a TTS service")
            TextToSpeech.ERROR_SYNTHESIS -> promise.reject("synthesis_error", "Failure of a TTS engine to synthesize the given input")
            TextToSpeech.LANG_MISSING_DATA -> promise.reject("lang_missing_data", "Language data is missing")
            TextToSpeech.LANG_NOT_SUPPORTED -> promise.reject("lang_not_supported", "Language is not supported")
            else -> promise.reject("error", "Unknown error code: $statusCode")
        }
    }

    private fun isPackageInstalled(packageName: String): Boolean {
        val pm = reactApplicationContext.packageManager
        return try {
            pm.getPackageInfo(packageName, 0)
            true
        } catch (e: PackageManager.NameNotFoundException) {
            false
        }
    }

    override fun getName(): String {
        return "TextToSpeech"
    }

    @ReactMethod
    fun getInitStatus(promise: Promise) {
        synchronized(initStatusPromises) {
            if (ready == null) {
                initStatusPromises.add(promise)
            } else {
                resolveReadyPromise(promise)
            }
        }
    }

    @ReactMethod
    fun speak(utterance: String, params: ReadableMap, promise: Promise) {
        if (notReady(promise)) return

        if (ducking) {
            val amResult = audioManager.requestAudioFocus(
                afChangeListener,
                AudioManager.STREAM_MUSIC,
                AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK
            )

            if (amResult != AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
                promise.reject("Android AudioManager error, failed to request audio focus")
                return
            }
        }

        val utteranceId = utterance.hashCode().toString()

        val speakResult = speak(utterance, utteranceId, params)
        if (speakResult == TextToSpeech.SUCCESS) {
            promise.resolve(utteranceId)
        } else {
            resolvePromiseWithStatusCode(speakResult, promise)
        }
    }

    @ReactMethod
    fun setDefaultLanguage(language: String, promise: Promise) {
        if (notReady(promise)) return

        val locale = if (language.contains("-")) {
            val parts = language.split("-")
            Locale(parts[0], parts[1])
        } else {
            Locale(language)
        }

        try {
            val result = tts?.setLanguage(locale)
            resolvePromiseWithStatusCode(result ?: -1, promise)
        } catch (e: Exception) {
            promise.reject("error", "Unknown error code")
        }
    }

    @ReactMethod
    fun setDucking(ducking: Boolean, promise: Promise) {
        if (notReady(promise)) return
        this.ducking = ducking
        promise.resolve("success")
    }

    @ReactMethod
    fun setDefaultRate(rate: Float, skipTransform: Boolean, promise: Promise) {
        if (notReady(promise)) return

        if (skipTransform) {
            val result = tts?.setSpeechRate(rate)
            resolvePromiseWithStatusCode(result ?: -1, promise)
        } else {
            val androidRate = if (rate < 0.5f) rate * 2 else rate * 4 - 1
            val result = tts?.setSpeechRate(androidRate)
            resolvePromiseWithStatusCode(result ?: -1, promise)
        }
    }

    @ReactMethod
    fun setDefaultPitch(pitch: Float, promise: Promise) {
        if (notReady(promise)) return
        val result = tts?.setPitch(pitch)
        resolvePromiseWithStatusCode(result ?: -1, promise)
    }

    @ReactMethod
    fun setDefaultVoice(voiceId: String, promise: Promise) {
        if (notReady(promise)) return

        if (Build.VERSION.SDK_INT >= 21) {
            try {
                tts?.voices?.forEach { voice ->
                    if (voice.name == voiceId) {
                        val result = tts?.setVoice(voice)
                        resolvePromiseWithStatusCode(result ?: -1, promise)
                        return
                    }
                }
            } catch (e: Exception) {
                // Purposefully ignore exceptions here due to some buggy TTS engines.
                // See http://stackoverflow.com/questions/26730082/illegalargumentexception-invalid-int-os-with-samsung-tts
            }
            promise.reject("not_found", "The selected voice was not found")
        } else {
            promise.reject("not_available", "Android API 21 level or higher is required")
        }
    }

    @ReactMethod
    fun voices(promise: Promise) {
        if (notReady(promise)) return

        val voiceArray = Arguments.createArray()

        if (Build.VERSION.SDK_INT >= 21) {
            try {
                tts?.voices?.forEach { voice ->
                    val voiceMap = Arguments.createMap().apply {
                        putString("id", voice.name)
                        putString("name", voice.name)

                        var language = iso3LanguageCodeToIso2LanguageCode(voice.locale.isO3Language)
                        val country = voice.locale.isO3Country
                        if (country.isNotEmpty()) {
                            language += "-${iso3CountryCodeToIso2CountryCode(country)}"
                        }

                        putString("language", language)
                        putInt("quality", voice.quality)
                        putInt("latency", voice.latency)
                        putBoolean("networkConnectionRequired", voice.isNetworkConnectionRequired)
                        putBoolean("notInstalled", voice.features.contains(TextToSpeech.Engine.KEY_FEATURE_NOT_INSTALLED))
                    }
                    voiceArray.pushMap(voiceMap)
                }
            } catch (e: Exception) {
                // Purposefully ignore exceptions here due to some buggy TTS engines.
                // See http://stackoverflow.com/questions/26730082/illegalargumentexception-invalid-int-os-with-samsung-tts
            }
        }

        promise.resolve(voiceArray)
    }

    @ReactMethod
    fun setDefaultEngine(engineName: String, promise: Promise) {
        if (notReady(promise)) return

        if (isPackageInstalled(engineName)) {
            ready = null
            onCatalystInstanceDestroy()
            tts = TextToSpeech(reactApplicationContext) { status ->
                synchronized(initStatusPromises) {
                    ready = (status == TextToSpeech.SUCCESS)
                    initStatusPromises.forEach { resolveReadyPromise(it) }
                    initStatusPromises.clear()
                    promise.resolve(ready)
                }
            }

            setUtteranceProgress()
        } else {
            promise.reject("not_found", "The selected engine was not found")
        }
    }

    @ReactMethod
    fun engines(promise: Promise) {
        if (notReady(promise)) return

        val engineArray = Arguments.createArray()

        if (Build.VERSION.SDK_INT >= 14) {
            try {
                val defaultEngineName = tts?.defaultEngine
                tts?.engines?.forEach { engine ->
                    val engineMap = Arguments.createMap().apply {
                        putString("name", engine.name)
                        putString("label", engine.label)
                        putBoolean("default", engine.name == defaultEngineName)
                        putInt("icon", engine.icon)
                    }
                    engineArray.pushMap(engineMap)
                }
            } catch (e: Exception) {
                promise.reject("error", "Unknown error code")
            }
        }

        promise.resolve(engineArray)
    }

    @ReactMethod
    fun stop(promise: Promise) {
        if (notReady(promise)) return

        val result = tts?.stop() ?: TextToSpeech.ERROR
        val resultValue = result == TextToSpeech.SUCCESS
        promise.resolve(resultValue)
    }

    @ReactMethod
    private fun requestInstallEngine(promise: Promise) {
        val intent = Intent(Intent.ACTION_VIEW).apply {
            data = Uri.parse("market://details?id=com.google.android.tts")
        }
        try {
            currentActivity?.startActivity(intent)
            promise.resolve("success")
        } catch (e: Exception) {
            promise.reject("error", "Could not open Google Text to Speech App in the Play Store")
        }
    }

    @ReactMethod
    private fun requestInstallData(promise: Promise) {
        val intent = Intent().apply {
            action = TextToSpeech.Engine.ACTION_INSTALL_TTS_DATA
        }
        try {
            currentActivity?.startActivity(intent)
            promise.resolve("success")
        } catch (e: ActivityNotFoundException) {
            promise.reject("no_engine", "No TTS engine installed")
        }
    }

    override fun onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy()
        tts?.apply {
            stop()
            shutdown()
        }
    }

    private fun notReady(promise: Promise): Boolean {
        return when {
            ready == null -> {
                promise.reject("not_ready", "TTS is not ready")
                true
            }
            ready != true -> {
                resolveReadyPromise(promise)
                true
            }
            else -> false
        }
    }

    @Suppress("DEPRECATION")
    private fun speak(utterance: String, utteranceId: String, inputParams: ReadableMap): Int {
        val audioStreamTypeString = inputParams.getString("KEY_PARAM_STREAM") ?: ""
        val volume = inputParams.getDouble("KEY_PARAM_VOLUME").toFloat()
        val pan = inputParams.getDouble("KEY_PARAM_PAN").toFloat()

        val audioStreamType = when (audioStreamTypeString) {
            "STREAM_ALARM" -> AudioManager.STREAM_ALARM
            "STREAM_DTMF" -> AudioManager.STREAM_DTMF
            "STREAM_MUSIC" -> AudioManager.STREAM_MUSIC
            "STREAM_NOTIFICATION" -> AudioManager.STREAM_NOTIFICATION
            "STREAM_RING" -> AudioManager.STREAM_RING
            "STREAM_SYSTEM" -> AudioManager.STREAM_SYSTEM
            "STREAM_VOICE_CALL" -> AudioManager.STREAM_VOICE_CALL
            else -> AudioManager.USE_DEFAULT_STREAM_TYPE
        }

        return if (Build.VERSION.SDK_INT >= 21) {
            val params = Bundle().apply {
                putInt(TextToSpeech.Engine.KEY_PARAM_STREAM, audioStreamType)
                putFloat(TextToSpeech.Engine.KEY_PARAM_VOLUME, volume)
                putFloat(TextToSpeech.Engine.KEY_PARAM_PAN, pan)
            }
            tts?.speak(utterance, TextToSpeech.QUEUE_ADD, params, utteranceId) ?: TextToSpeech.ERROR
        } else {
            val params = HashMap<String, String>().apply {
                put(TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID, utteranceId)
                put(TextToSpeech.Engine.KEY_PARAM_STREAM, audioStreamType.toString())
                put(TextToSpeech.Engine.KEY_PARAM_VOLUME, volume.toString())
                put(TextToSpeech.Engine.KEY_PARAM_PAN, pan.toString())
            }
            tts?.speak(utterance, TextToSpeech.QUEUE_ADD, params) ?: TextToSpeech.ERROR
        }
    }

    private fun sendEvent(eventName: String, utteranceId: String) {
        val params = Arguments.createMap().apply {
            putString("utteranceId", utteranceId)
        }
        sendEvent(eventName, params)
    }

    private fun sendEvent(eventName: String, params: WritableMap) {
        reactApplicationContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(eventName, params)
    }

    /**
     * Update to React Native 0.74.2
     */
    @ReactMethod
    fun removeListeners(count: Int) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    @ReactMethod
    fun addListener(eventName: String) {
        // Implement your logic to add the listener
        // You can store the handler in a Map or List if necessary
    }

    @ReactMethod
    fun removeListener(eventName: String) {
        // Implement your logic to remove the listener
        // You can remove the handler from the Map or List if necessary
    }
}
